File: simpletranscribe-win/SimpleTranscribe/Models/DownloadProgress.cs
namespace SimpleTranscribe.Models;

public record DownloadProgress(
    double Fraction,
    double BytesPerSecond,
    long TotalBytes,
    long ReceivedBytes)
{
    public double? EtaSeconds =>
        TotalBytes > 0 && BytesPerSecond > 0
            ? (TotalBytes - ReceivedBytes) / BytesPerSecond
            : null;

    public string SpeedString
    {
        get
        {
            double mbs = BytesPerSecond / 1_048_576;
            if (mbs >= 1)
                return $"{mbs:F1} MB/s";

            double kbs = BytesPerSecond / 1_024;
            return $"{Math.Max(kbs, 0):F0} KB/s";
        }
    }

    public string EtaString
    {
        get
        {
            if (EtaSeconds is not { } eta)
                return "";
            if (eta < 60)
                return $"~{eta:F0}s remaining";
            if (eta < 3600)
                return $"~{eta / 60:F0}min remaining";
            return $"~{eta / 3600:F1}h remaining";
        }
    }
}

File: simpletranscribe-mac/simpletranscribe/Models/DownloadProgress.swift
import Foundation

struct DownloadProgress {
    let fraction: Double
    let bytesPerSecond: Double
    let totalBytes: Int64
    let receivedBytes: Int64

    var etaSeconds: Double? {
        guard totalBytes > 0, bytesPerSecond > 0 else { return nil }
        let remaining = Double(max(totalBytes - receivedBytes, 0))
        return remaining / bytesPerSecond
    }

    var speedString: String {
        let mbs = bytesPerSecond / 1_048_576
        if mbs >= 1 {
            return String(format: "%.1f MB/s", mbs)
        }
        let kbs = bytesPerSecond / 1_024
        return String(format: "%.0f KB/s", max(kbs, 0))
    }

    var etaString: String {
        guard let eta = etaSeconds else { return "" }
        if eta < 60 {
            return String(format: "~%.0fs remaining", eta)
        }
        if eta < 3600 {
            return String(format: "~%.0fmin remaining", eta / 60)
        }
        return String(format: "~%.1fh remaining", eta / 3600)
    }
}
